Initial baseline

CAD97 opened the pull request:
This should be an uncontroversial subset of the string indexing behaviors.

I'm happy to give over control of the crate to either rust-analyzer and/or codespan once we've finalized the API, but I do hope we can push this out as 1.0 initially and then use/re-export it in both places.

Feel free to indicate what APIs are desired but not yet present on this PR review as well.

r? @matklad
r? @brendanzab

CAD97 committed:
Initial implementation

CAD97 commented:
For reference:

- `StrIndex`
  - `codespan`: https://docs.rs/codespan/0.5.0/codespan/struct.ByteIndex.html
  - `codespan`: https://docs.rs/codespan/0.5.0/codespan/struct.ByteOffset.html
  - `text_unit`: https://docs.rs/text_unit/0.1.9/text_unit/struct.TextUnit.html
  - `codemap`: https://docs.rs/codemap/0.1.2/codemap/struct.Pos.html
- `StrRange`
  - `codespan`: https://docs.rs/codespan/0.5.0/codespan/struct.Span.html
  - `text_unit`: https://docs.rs/text_unit/0.1.9/text_unit/struct.TextRange.html
  - `codemap`: https://docs.rs/codemap/0.1.2/codemap/struct.Span.html
  - `language-reporting`: https://docs.rs/language-reporting/0.3.1/language_reporting/trait.ReportingSpan.html
  - `language-reporting`: https://docs.rs/language-reporting/0.3.1/language_reporting/struct.SimpleSpan.html
  - `annotate-snippets`: https://docs.rs/annotate-snippets/0.6.1/annotate_snippets/snippet/struct.Slice.html

CAD97 committed:
Length of a range

matklad commented on Cargo.toml:
```diff
@@ -9,6 +9,5 @@ authors = [
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
 [dependencies]
+serde = { version = "1.0", optional = true, features = ["derive"] }
```
I wonder if implementing serde manually worth the improvement in compile times. It’s seems great If vocabulary crates don’t need to wait for syn to be build. OTOH, probability that all the crates use serde without derive feature seems small.

Hmm, actually, derived deserialize can violate start < end contract, so we better implement it manually anyway

CAD97 replied:
I hope you like manual deserialization code because that's a lot of it added to remove the derive dependency!

Normally, one would implement deserialize manually by deserializing a helper and then validating on that, but I do see the value in handwriting it to skip the derive dependency in a small leaf crate with only a couple types.

matklad commented on src/fmt.rs:
```diff
@@ -0,0 +1,26 @@
+
+impl fmt::Debug for StrIndex {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self)
```
Will this handle padding and alignment automatically? Perhaps it”s better just to call fmt::Display::fmt(raw)?

CAD97 replied:
Yes, any writes in `Debug::fmt` are already padded/aligned; you go through the `f.debug_*` functions for something more complicated than a primitive. This `write!` call is equivalent to manually calling `Display::fmt`, but it's slightly more idiomatic to go through the `write!` call. The `format_args!` machinery is specifically optimized to disappear completely in the just-forward-one-`fmt`-call case.

CAD97 replied:
I decided to go ahead and make the forwards-to-`Display` obvious by just using `fmt::Display::fmt`.

matklad commented on src/lib.rs:
````diff
@@ -0,0 +1,109 @@
+    ///     StrRange::between(10.into(), 20.into()),
+    /// );
+    /// ```
+    pub fn from_len(start: StrIndex, len: StrIndex) -> Self {
````
That”s a name i want to change in text unit. Originally it was created for the symmetry with from_to, but, while from_to reads unambiguously, from_len can be read in two ways (as “from, len” and as “from this len”)

CAD97 replied:
Would it make sense to spell this like `StrRange::from(index).with_len(len)`? That's the only other reasonable spelling I can come up with.

CAD97 replied:
I've decided to remove this function from this PR and add it in a future PR.

matklad commented on src/ops.rs:
```diff
@@ -0,0 +1,52 @@
+}
+
+math!(impl Add for StrIndex by fn add = +);
+math!(impl Sub for StrIndex by fn sub = -);
```
Nice syntax!

CAD97 committed:
Combinations of ranges

CAD97 committed:
Make Debug->Display forwarding clearer

CAD97 committed:
Remove StrRange::from_len for now

A different spelling is planned,
but this PR#1 is supposed to be the "obvious" API surface

CAD97 committed:
Add examples for StrRange::between

Specifically, to test the construction of an empty unit range.

matklad commented on src/fmt.rs:
```diff
@@ -0,0 +1,28 @@
+
+impl fmt::Display for StrIndex {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.raw)
```
Alignment is still lost here

CAD97 replied:
What do you mean? [This is works fine as far as I can tell.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0ed5ae7f08fc42ef2aefef337a4aa164)

matklad replied:
See https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f60485ee5d2f7c58440388bc764cc174

CAD97 replied:
Ah ok.

matklad commented on src/serde.rs:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+impl Serialize for StrRange {
```
Hm, maybe serialize this as newtype warapper of a tupple? Should make JSON more compact and readable.

CAD97 replied:
Personally, I think I slightly prefer `{ "start": 0, "end": 10 }` to `[0, 10]` (especially since that looks like `[0..=10]` rather than `[0..10)`!), though I understand that requires manually going and adjusting the pretty JSON formatting.

Also, mostly note to self but also readers, the serialization format does not mean it has to be represented in memory that way. We could switch to `{ start: u32, len: u16 }` and still serialize as `{ start: u32, end: u32 }` just fine, if we so desired.

brendanzab commented:
Might want to get @Marwes to look at this too!

brendanzab commented:
I guess the issue with trying to standardise on a single type is that there might be project-specific concerns for data packing and compression. Not sure whether this is something that could be integrated into a centralised library though - perhaps it could.

brendanzab commented on src/lib.rs:
````diff
@@ -1 +1,109 @@
+}
+
+impl StrIndex {
+    /// Index equal to the string length of this `char`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use str_index::*;
+    /// assert_eq!(
+    ///     StrIndex::from_char_len('😂'),
+    ///     StrIndex::from(4),
+    /// );
+    /// ```
+    pub fn from_char_len(c: char) -> Self {
+        StrIndex::from(c.len_utf8() as u32)
+    }
````
Might be worth having a utf16 version of this?

CAD97 replied:
I think the correct place to do the Rust utf8 str indexes to utf16 indexes is at the language server protocol edge. Especially since `StrIndex` is supposed to be able to index into `str`.

CAD97 commented:
On packing: 

Something like `SourceMap` works by mapping all individual files' source into one address space. This is still compatible with the one-dimensional `StrIndex`.

`StrRange` compression is typically done by interning to a smaller key. I suppose if you're willing to limit any single file to <64 KiB, you could represent `StrRange` as `(start: u32, len: u16)` to save a byte per span, but that seems overly restrictive for a vocabulary type. I don't think anyone currently does that, though; cross-virtual-file spans do exist. If I were implementing `SourceMap` from scratch without outside influence I'd probably make a 32 bit location `(file: u16, pos: u16)` and allow one physical file to span multiple `SourceMap` "file"s. TL;DR bring your own `SourceMap`/`FileDB` and map to that domain yourself.

Marwes commented:
I don't really see this crate carrying its own weight (at least if it is this bare bones). As it is there are only a newtype over `u32` and a standard range type `(u32, u32)`.

The newtype is arguably useful in that it explains that the index is into a string but without anything else to build on it then each user could just specify that themselves. To actually use it (ie index into the string it is associated to) then it needs to be converted to `usize` anyway so there isn't any difference in how it is used from a unique one per user.

The span could be useful with some extra funcions built on it (like codespan used to have) but otherwise it is just a https://doc.rust-lang.org/std/ops/struct.Range.html, so why not build on that as the standard type instead?

CAD97 commented:
I intend(ed) to implement `Index[Mut]<StrRange> for str` and `Index[Mut]<impl ops::RangeBounds<StrIndex>> for str`, the only reason they aren't in this PR is I forgot until now.

These types are _intended_ to be vocabulary types used to communicate between disparate compiler library crates (though that depends on adoption of course, this is just a sketch proposal currently).

I originally made `StrRange` "just" `std::ops::Range<StrIndex>`, but having a unique type gives tangible benefits: `Copy` (by not being an iterator) and an ordering guarantee (e.g. `10..0` is not valid).

CAD97 commented:
Here's another new reason to use a dedicated `StrRange` and not `ops::Range<StrIndex>`: for the life of me I cannot make `<str as Index<ops::Range<StrIndex>>>` true. I can't even get there via `#![feature(slice_index_methods)]` because `SliceIndex` is properly sealed via the priv-in-pub-supertrait trick. (It did involve a [nice tt-muncher decl-macro](https://users.rust-lang.org/t/tt-muncher-into-function-args/33447) that's now useless, though.)

CAD97 committed:
Add indexing capability for &str[StrRange]

CAD97 committed:
Initialize StrRange using From<ops::Range>

CAD97 commented on src/lib.rs:
````diff
@@ -1 +1,109 @@
+    ///
+    /// StrRange::between(0.into(), 0.into());
+    /// ```
+    pub fn between(start: StrIndex, end: StrIndex) -> Self {
````
Note that #9 proposes an alternative construction method of `StrRange::from(start..end)` (and `start..=end`, and `..end`, etc.) instead of `StrRange::between(start, end)`.

CAD97 committed:
Allow `StrIndex` to be formatted with alignment
by forwarding directly to raw implementation of the fmt traits

CAD97 committed:
Fix indexing tests

CAD97 committed:
Features: don't enable `serde/alloc` from `alloc`

That means `alloc` implies `serde`.
We don't use it anyway.
